internal/controller/dns/record.go

```diff
@@ -0,0 +1,223 @@
+	}
+
+	if cr.Spec.ForProvider.Zone == nil {
+		return managed.ExternalObservation{}, errors.New(errDNSRecordNoZone)
+	}
+
+	record, err := e.client.DNSRecord(ctx, *cr.Spec.ForProvider.Zone, en)
+
+	if err != nil {
+		// Been deleted or doesnt exist
+		if records.IsRecordNotFound(err) {
+			return managed.ExternalObservation{ResourceExists: false}, nil
+		}
+		return managed.ExternalObservation{}, errors.Wrap(err, errDNSRecordLookup)
+	}
+
+	cr.Status.AtProvider = records.GenerateObservation(record)
+
+	return managed.ExternalObservation{
+		ResourceExists:          true,
+		ResourceUpToDate:        records.UpToDate(&cr.Spec.ForProvider, record),
+		ResourceLateInitialized: records.LateInitialize(&cr.Spec.ForProvider, record),
+	}, nil
+}
+
+func (e *external) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mg.(*v1alpha1.Record)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotDNSRecord)
+	}
+
+	if cr.Spec.ForProvider.Zone == nil {
+		return managed.ExternalCreation{}, errors.New(errDNSRecordCreation)
+	}
+
+	if cr.Spec.ForProvider.TTL == nil {
+		return managed.ExternalCreation{}, errors.New(errDNSRecordCreation)
+	}
+
+	// TODO: Add validation here for priority (only required for specific record types)
+
+	cr.SetConditions(rtv1.Creating())
+
+	res, err := e.client.CreateDNSRecord(
+		ctx,
+		*cr.Spec.ForProvider.Zone,
+		cloudflare.DNSRecord{
+			Type:     *cr.Spec.ForProvider.Type,
+			Name:     cr.Spec.ForProvider.Name,
+			TTL:      *cr.Spec.ForProvider.TTL,
+			Content:  cr.Spec.ForProvider.Content,
+			Proxied:  cr.Spec.ForProvider.Proxied,
+			Priority: cr.Spec.ForProvider.Priority,
+		},
+	)
+
+	if err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, errDNSRecordCreation)
+	}
+
+	cr.Status.AtProvider = records.GenerateObservation(res.Result)
+
+	// Update the external name with the ID of the new Zone
+	meta.SetExternalName(cr, res.Result.ID)
+
+	return managed.ExternalCreation{ExternalNameAssigned: true}, nil
+}
+
+func (e *external) Update(ctx context.Context, mg resource.Managed) (managed.ExternalUpdate, error) {
+	cr, ok := mg.(*v1alpha1.Record)
+	if !ok {
+		return managed.ExternalUpdate{}, errors.New(errNotDNSRecord)
+	}
+
+	if cr.Spec.ForProvider.Zone == nil {
+		return managed.ExternalUpdate{}, errors.New(errDNSRecordDeletion)
+	}
+
+	return managed.ExternalUpdate{},
+		errors.Wrap(
+			records.UpdateRecord(ctx, e.client, meta.GetExternalName(cr), &cr.Spec.ForProvider),
+			errDNSRecordUpdate,
+		)
+}
+
+func (e *external) Delete(ctx context.Context, mg resource.Managed) error {
+	cr, ok := mg.(*v1alpha1.Record)
+	if !ok {
+		return errors.New(errNotDNSRecord)
+	}
+
+	if cr.Spec.ForProvider.Zone == nil {
+		return errors.New(errDNSRecordDeletion)
+	}
+
+	err := e.client.DeleteDNSRecord(ctx, *cr.Spec.ForProvider.Zone, meta.GetExternalName(cr))
+
+	return errors.Wrap(err, errDNSRecordDeletion)
```

Comment:
No need to allocate the `err` var here, just put the `DeleteDNSRecord` call inside `errors.Wrap`.